File: tests/test_bdpt_config.cpp
#include "division/BdptConfig.h"

#include <cassert>
#include <string>
#include <vector>

int main() {
    {
        BdptRunConfig config;
        std::string error;
        assert(parseBdptOptionalArgs({}, config, error));
        assert(config.crossMode == BdptCrossMode::Exact);
        assert(config.unitSearchMode == BdptUnitSearchMode::MinPin);
        assert(config.signLabeling);
        assert(!config.reproduction);
        assert(config.timerSeconds == 86400);
        assert(config.threads == 8);
    }

    {
        BdptRunConfig config;
        std::string error;
        const std::vector<std::string> args = {
            "cross", "paper",
            "solver", "per-bit",
            "sign", "1",
            "repro", "1",
            "timer", "600",
            "threads", "4",
        };
        assert(parseBdptOptionalArgs(args, config, error));
        assert(config.crossMode == BdptCrossMode::Paper);
        assert(config.unitSearchMode == BdptUnitSearchMode::PerBit);
        assert(config.signLabeling);
        assert(config.reproduction);
        assert(config.timerSeconds == 600);
        assert(config.threads == 4);
        assert(toString(config.crossMode) == "paper");
        assert(toString(config.unitSearchMode) == "per-bit");
    }

    {
        BdptRunConfig config;
        std::string error;
        assert(!parseBdptOptionalArgs({"cross", "relaxed"}, config, error));
        assert(error.find("cross") != std::string::npos);
    }

    {
        BdptRunConfig config;
        std::string error;
        assert(!parseBdptOptionalArgs({"solver", "per-bit", "repro"}, config, error));
        assert(error.find("pairs") != std::string::npos);
    }

    {
        BdptRunConfig config;
        std::string error;
        assert(!parseBdptOptionalArgs({"unknown", "1"}, config, error));
        assert(error.find("unknown") != std::string::npos);
    }

    {
        BdptRunConfig config;
        std::string error;
        assert(!parseBdptOptionalArgs(
            {"repro", "1", "sign", "0"}, config, error));
        assert(error.find("sign 1") != std::string::npos);
    }

    return 0;
}

File: experiments/test_bdpt_config.py
"""Regression tests for -div3 BDPT optional argument parsing."""

from __future__ import annotations

import shutil
import subprocess
import tempfile
import textwrap
import unittest
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]


class BdptConfigParserTests(unittest.TestCase):
    @unittest.skipIf(shutil.which("g++") is None, "g++ is required for this parser test")
    def test_strict_reproduction_options_are_parseable(self) -> None:
        source = textwrap.dedent(
            r"""
            #include "division/BdptConfig.h"

            #include <iostream>
            #include <string>
            #include <vector>

            int main() {
                BdptRunConfig config;
                std::string error;
                std::vector<std::string> args = {
                    "cross", "paper",
                    "solver", "per-bit",
                    "sign", "1",
                    "repro", "1",
                    "timer", "42",
                    "threads", "3",
                };

                if (!parseBdptOptionalArgs(args, config, error)) {
                    std::cerr << error << "\n";
                    return 1;
                }
                if (config.crossMode != BdptCrossMode::Paper) return 2;
                if (config.unitSearchMode != BdptUnitSearchMode::PerBit) return 3;
                if (!config.signLabeling) return 4;
                if (!config.reproduction) return 5;
                if (config.timerSeconds != 42) return 6;
                if (config.threads != 3) return 7;
                return 0;
            }
            """
        )

        with tempfile.TemporaryDirectory() as td:
            test_cpp = Path(td) / "bdpt_config_test.cpp"
            exe = Path(td) / "bdpt_config_test.exe"
            test_cpp.write_text(source, encoding="utf-8")

            compile_cmd = [
                "g++",
                "-std=c++14",
                "-I",
                str(ROOT / "include"),
                str(test_cpp),
                str(ROOT / "lib" / "division" / "BdptConfig.cpp"),
                "-o",
                str(exe),
            ]
            compiled = subprocess.run(
                compile_cmd,
                cwd=ROOT,
                text=True,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                check=False,
            )
            self.assertEqual(
                compiled.returncode,
                0,
                msg=compiled.stdout + compiled.stderr,
            )

            completed = subprocess.run(
                [str(exe)],
                cwd=ROOT,
                text=True,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                check=False,
            )
            self.assertEqual(
                completed.returncode,
                0,
                msg=completed.stdout + completed.stderr,
            )


if __name__ == "__main__":
    unittest.main()
